generate/unmarshal.go.tmpl

```diff
@@ -1,40 +1,40 @@
 {{/* (the blank lines at the start are intentional, to separate
      UnmarshalJSON from the function it follows) */}}
 
-func (v *{{.Type}}) UnmarshalJSON(b []byte) error {
-	var firstPass struct{
-		*{{.Type}}
-        {{range .Fields -}}
-		{{.GoName}} {{ref "encoding/json.RawMessage"}} `json:"{{.JSONName}}"`
-        {{end}}
-	}
-	firstPass.{{.Type}} = v
+func (v *{{.GoName}}) UnmarshalJSON(b []byte) error {
+    {{/* We want to specially handle the abstract fields (.AbstractFields),
+         but unmarshal everything else normally.  To handle abstract fields,
+         first we unmarshal them into a json.RawMessage, and then handle those
+         further, below.  For the rest, we just want to call json.Unmarshal.
+         But if we do that naively on a value of type `.Type`, it will call
+         this function again, and recurse infinitely.  So we make a wrapper
+         type -- with a different name, thus different methods, but the same
+         fields, and unmarshal into that.  For more on why this is so
+         difficult, see
+         https://github.com/benjaminjkraft/notes/blob/master/go-json-interfaces.md
```

Comment:
Thanks for including this reference! The unmarshaling is definitely tricky, but it makes sense.